### ISSUE TITLE
provide opt-out var to use public IPv4 addresses only

### ISSUE THREAD

**Issue by nusenu** [nusenu/ansible-relayor#91]
We filter IPv6 addresses for public IPs but not on IPv4 yet.

```
ipaddr('public') 
```

**Comment by nusenu** [comment]
there might be still environments where this would be needed? (server with public and non-public addresses) I guess this is rarely the case.

**Comment by kberanek** [comment]
>  (server with public and non-public addresses) I guess this is rarely the case.

This is exactly how a number of my relays are configured.  I only have a single public IP for those relays so I simply remove the Address-related fields in my torrc template: https://github.com/kberanek/ansible-relayor/blob/mine/templates/torrc

This obviously doesn't work if you have multiple public IPs.

**Comment by nusenu** [comment]
Ok, so we want a var that allows to opt-out of using RFC1918 IPs and a check that aborts if we are left with no IPs.

Why did you touch ORPort, DirPort and OutboundBindAddress as well? (this will cause trouble on relays with more than one public IP)
https://github.com/kberanek/ansible-relayor/commit/9d3c53a35d1e88e29c1ce0740bf45a08e69b550d

**Comment by nusenu** [comment]
testing welcome.

if you set 
```tor_publicIPsOnly: True ```
with no public IPs, it will fail when writing the torrc config file.

**Comment by kberanek** [comment]
I don't think there was any reason to touch ORPort, DirPort and OutboundBindAddress.  Not sure why I did that.

I tested https://github.com/kberanek/ansible-relayor/commit/8f1ecb58f0c1c69fce8ef8f36776e468180129f3 with tor_publicIPsOnly set to first True and then False on a machine that only knows about a private IP (a 10/8 address) and with it set to False everything worked as expected, but when I set it to true it doesn't really "fail".  If I 

```
TASK [ansible-relayor : setup IP list (2/2)] ***********************************
task path: /home/kevin/src/relay-setup/ansible-relayor/tasks/ip-list.yml:30
ok: [tr2] => {
    "ansible_facts": {
        "tor_ips": []
    }, 
    "changed": false, 
    "invocation": {
        "module_args": {
            "tor_ips": []
        }, 
        "module_name": "set_fact"
    }
}

....


PLAY RECAP *********************************************************************
tr2                        : ok=18   changed=4    unreachable=0    failed=0 
```

**Comment by kberanek** [comment]
Maybe that's at least partly due to running the test on a machine that isn't a clean install.  I could try it out on a totally clean machine, but probably not until tomorrow.

**Comment by nusenu** [comment]
kberanek@8f1ecb5 is not expected to fail.
e05f6f47e9077758337b16b0d378ca5fe6b6d6e7 is expected to fail.

**Comment by kberanek** [comment]
Understood.  The 8f1ecb5 commit is a cherry-pick of your e05f6f4 commit on
top of my changes. I ran it with the flag set to both true and false, but
it looks like the version I pushed to that branch happened to have it set
to false.  When I ran it with the flag set to true, I just saw an empty
`tor_ips`, which you can see in the excerpt of the logs I pasted
previously. It looked like the only real effect was that all the tasks that
iterate over instances using `with_nested` we're just turned into no-ops
because `tor_ips` was empty.

**Comment by nusenu** [comment]
Yes ```tor_ips``` is empty, but tor_ports isn't.

It should fail at 
```Generating torrc file(s)```
specifically due to the 
```validate: "tor --verify-config -f %s"``` (it does in my case when ```tor_ips``` is []).

but I **might** add a task to check 
```tor_ips``` for []

**Comment by kberanek** [comment]
It looks like there's something in my fork and/or playbook that causes this to not behave the same way.  Not sure exactly that is yet, but if I use https://github.com/nusenu/ansible-relayor/commit/e05f6f47e9077758337b16b0d378ca5fe6b6d6e7 with a trivial playbook (see below), I can confirm that it does indeed fail as expected.

The only thing that seems a bit weird is that before failing it will generate users with names like `_tor-_443` and create directories like `/var/lib/tor-instances/_443`.  Failing relatively early on if `tor_ips` is empty would prevent this from happening.  It looks like it does fail before actually installing these instances as system services (creating systemd unit files, etc.) so it's not really a big deal, but it seems to me like it would be ideal if it didn't create these things.

```
---

- hosts: relays
  vars:
    - tor_publicIPsOnly: True
  roles:
    - ansible-relayor
```

**Comment by kberanek** [comment]
Also, I rediscovered the reason why I removed the address from ORPort and DirPort before.  If I don't do that I end up with logs that look like the one below.  I played around with some approaches to generating a torrc that correctly uses both the internal and external IPs, but I didn't really like anything I came up with.  Given that I don't run any relays with multiple IPs, removing the references to the internal address seemed like the easiest solution for me.  

```
Jan 21 19:20:39 tr4 Tor-10.8.169.135_443[8258]: The IPv4 DirPort address 10.8.169.135 does not match the descriptor address 51.15.48.254. If you have a static public IPv4 address, use 'Address <IPv4>' and 'OutboundBindAddress <IPv4>'. If you are behind a NAT, use two DirPort lines: 'DirPort <PublicPort> NoListen' and 'DirPort <InternalPort> NoAdvertise'.
```

**Comment by nusenu** [comment]
Kevin Beranek:
> but it seems to me like it would be ideal if it didn't create these things.

I agree, thank you for your input!

**Comment by nusenu** [comment]
If the server has public and private IPv4 addresses (all listed in ansible_all_ipv4_addresses) we currently use all of them, which is not what we want if the private IP gets NATed to the public IP. In such an (unlikely) environment we would create 4 instances with a single public IPv4 address only - which is bad.

to avoid this:
use exclusively public IPv4 addresses by default, if we have no public IPv4 address use a single private IPv4 address.

optional: 
provide a var to use all private IPv4 addresses (in that case the operator has to make sure that every private IPv4 address has a public IPv4 address it gets NATed to).

**Comment by nusenu** [comment]
implemented in 93a6b1e6f8cfd153ab88b25d6e9880758630493b.

Testing and feedback welcom.

**Comment by nusenu** [comment]
superseded by #102
